airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_with_global.py

```diff
@@ -0,0 +1,179 @@
+
+    - Uses the `_get_active_cursor()` helper method to select the active cursor based on the `_use_global_cursor` flag.
+    - This simplifies the logic and ensures consistent cursor usage across methods.
+
+    **State Structure Example**
+
+    ```json
+    {
+        "states": [
+            {
+                "partition": {"partition_key": "partition_1"},
+                "cursor": {"cursor_field": "2021-01-15"}
+            },
+            {
+                "partition": {"partition_key": "partition_2"},
+                "cursor": {"cursor_field": "2021-02-14"}
+            }
+        ],
+        "state": {
+            "cursor_field": "2021-02-15"
+        },
+        "use_global_cursor": false
+    }
+    ```
+
+    In this example, the cursor is using partition-based state management (`"use_global_cursor": false`), maintaining separate cursor states for each partition.
+
+    **Usage Scenario**
+
+    Suitable for streams where the number of partitions may vary significantly, requiring dynamic switching between per-partition and global state management to ensure data consistency and efficient synchronization.
+    """
+
+    def __init__(self, cursor_factory: CursorFactory, partition_router: PartitionRouter, stream_cursor: DatetimeBasedCursor):
+        self._partition_router = partition_router
+        self._per_partition_cursor = PerPartitionCursor(cursor_factory, partition_router)
+        self._global_cursor = GlobalSubstreamCursor(stream_cursor, partition_router)
+        self._use_global_cursor = False
+        self._current_partition: Optional[Mapping[str, Any]] = None
+        self._last_slice: bool = False
+
+    def _get_active_cursor(self) -> Union[PerPartitionCursor, GlobalSubstreamCursor]:
+        return self._global_cursor if self._use_global_cursor else self._per_partition_cursor
+
+    def stream_slices(self) -> Iterable[StreamSlice]:
+        self._global_cursor.start_slices_generation()
+
+        # Iterate through partitions and process slices
+        for partition, is_last_partition in iterate_with_last_flag(self._partition_router.stream_slices()):
+            # Generate slices for the current cursor and handle the last slice using the flag
+            self._current_partition = partition.partition
+            for slice, is_last_slice in iterate_with_last_flag(
+                self._get_active_cursor().generate_slices_from_partition(partition=partition)
+            ):
+                self._global_cursor.register_slice(is_last_slice and is_last_partition)
+                yield slice
+        self._current_partition = None
+        self._last_slice = True
+
+    def set_initial_state(self, stream_state: StreamState) -> None:
+        """
+        Set the initial state for the cursors.
+        """
+        self._use_global_cursor = stream_state.get("use_global_cursor", False)
+
+        self._global_cursor.set_initial_state(stream_state)
+        self._per_partition_cursor.set_initial_state(stream_state)
```

Comment:
Should this be done only if `not self._use_global_cursor`?

Reply:
Fixed. However, it doesn’t make much of a difference since [we don’t save the per-partition cursor](https://github.com/airbytehq/airbyte/pull/45125/files/1fb01a90c03c56dee72eff2acf77502ba6d858e0#diff-04f9f6f5abeb2f86ff177a3472801b8ca9c65aa39828d018281a17890b9efa54R115) if the global cursor is being used.